File: 0987.二叉树的垂序遍历/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func verticalTraversal(root *TreeNode) [][]int {
	type node struct {
		val, x, y int
	}
	vec := make([]node, 0)
	var dfs func(*TreeNode, int, int)
	dfs = func(root *TreeNode, x, y int) {
		if root == nil {
			return
		}
		vec = append(vec, node{root.Val, x, y})
		if root.Left != nil {
			dfs(root.Left, x+1, y-1)
		}
		if root.Right != nil {
			dfs(root.Right, x+1, y+1)
		}
	}
	dfs(root, 0, 0)
	sort.Slice(vec, func(a, b int) bool {
		return (vec[a].y < vec[b].y) ||
			(vec[a].y == vec[b].y && vec[a].x < vec[b].x) ||
			(vec[a].y == vec[b].y && vec[a].x == vec[b].x && vec[a].val < vec[b].val)
	})
	ans := make([][]int, 0)
	lst := math.MinInt32 / 2
	for _, i := range vec {
		val, y := i.val, i.y
		if y != lst {
			ans = append(ans, nil)
		}
		ans[len(ans)-1] = append(ans[len(ans)-1], val)
		lst = y
	}
	return ans
}

File: 0987.二叉树的垂序遍历/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
struct node{
    int val,x,y;
};
class Solution {
    vector<node> vec;
public:
    vector<vector<int>> verticalTraversal(TreeNode* root) {
        vec.clear();
        dfs(root,0,0);
        sort(vec.begin(),vec.end(),[](const node & A,const node & B){
            return (A.y<B.y) ||
                (A.y==B.y && A.x<B.x) ||
                (A.y==B.y && A.x==B.x && A.val<B.val);
        });
        vector<vector<int>> ans;
        int lst=INT_MIN/2;
        for (node &i:vec) {
            int val=i.val,y=i.y;
            if (y!=lst) {
                ans.push_back({});
            }
            lst=y;
            ans.back().push_back(val);
        }
        return ans;
    }
    void dfs(TreeNode *root,int x,int y) {
        if (root==NULL) {
            return;
        }
        vec.push_back({root->val,x,y});
        if (root->left!=NULL) {
            dfs(root->left,x+1,y-1);
        }
        if (root->right!=NULL) {
            dfs(root->right,x+1,y+1);
        }
    }
};